apps/framework-cli/src/utilities/validate_passthrough.rs

```diff
@@ -678,33 +678,143 @@ impl<'de, S: SerializeValue> Visitor<'de> for &mut ValueVisitor<'_, S> {
                     .serialize_value(&map)
                     .map_err(A::Error::custom)
             }
-            ColumnType::Json => {
-                struct JsonPassThrough<'de, MA: MapAccess<'de>> {
+            ColumnType::Json(opts) => {
+                struct JsonPassThrough<'de, 'a, MA: MapAccess<'de>> {
                     map: RefCell<MA>,
+                    opts: &'a crate::framework::core::infrastructure::table::JsonOptions,
+                    parent_context: &'a ParentContext<'a>,
                     _phantom_data: &'de PhantomData<()>,
                 }
-                impl<'de, MA: MapAccess<'de>> Serialize for JsonPassThrough<'de, MA> {
+                impl<'de, 'a, MA: MapAccess<'de>> Serialize for JsonPassThrough<'de, 'a, MA> {
                     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                     where
                         S: Serializer,
                     {
                         use serde::ser::Error;
+                        use std::collections::HashMap;
+
+                        // Group typed paths by their top-level key
+                        #[derive(Clone)]
+                        struct PathSpec {
+                            full: String,
+                            segments: Vec<String>,
+                            required: bool,
+                            seen: bool,
+                        }
+
+                        let mut by_top: HashMap<String, Vec<PathSpec>> = HashMap::new();
+                        for (p, t) in &self.opts.typed_paths {
+                            let required = !matches!(t, ColumnType::Nullable(_));
+                            let mut parts = p.split('.').map(|s| s.to_string());
+                            if let Some(top) = parts.next() {
+                                let rest: Vec<String> = parts.collect();
+                                by_top.entry(top).or_default().push(PathSpec {
+                                    full: p.clone(),
+                                    segments: rest,
+                                    required,
+                                    seen: false,
+                                });
+                            }
+                        }
+
                         let mut inner = serializer.serialize_map(None)?;
                         let mut map = self.map.borrow_mut();
                         while let Some(key) = map.next_key::<String>().map_err(S::Error::custom)? {
-                            inner.serialize_key(&key).map_err(S::Error::custom)?;
-                            SerializeMap::serialize_value(
-                                &mut inner,
-                                &map.next_value::<serde_json::Value>()
-                                    .map_err(S::Error::custom)?,
-                            )?;
+                            // Validate typed paths under this top-level key if any
```

Comment:
Looks like this method got way to big. Probably should be split into smaller chunks